src/Polysemy/Internal/TH/Effect.hs

```diff
@@ -75,135 +79,185 @@ makeSem = genFreer True
-      sigs <- filter (const makeSigs) <$> mapM genSig cons
-      decs <- join <$> mapM genDecl cons
-      loc <- location
-
-      return $
-        [ TySynInstD ''DefiningModule
-            . TySynEqn [ConT tcName]
-            . LitT
-            . StrTyLit
-            $ loc_module loc
-        | hasTyFams
-        ] ++ sigs ++ decs
-
-    _ -> fail "makeSem expects a type constructor"
-
--- | Given the name of a GADT constructor, return the name of the corresponding
--- lifted function.
-getDeclName :: Name -> Name
-getDeclName = mkName . overFirst toLower . nameBase
- where
-  overFirst f (a : as) = f a : as
-  overFirst _ as       = as
+genFreer shouldMkSigs typeName = do
+  dtInfo <- reifyDatatype typeName
+  -- TODO: data families seem to work just fine - should they be restricted?
+  clInfos  <- traverse (mkCLInfo dtInfo) $ datatypeCons dtInfo
+  defModFI <- TySynInstD ''DefiningModule
+          .   TySynEqn [ConT $ datatypeName dtInfo]
+          .   LitT
+          .   StrTyLit
+          .   loc_module
+          <$> location
+  sigs     <- if shouldMkSigs then traverse genSig clInfos
+                              else return []
+  decs     <- traverse genDec clInfos
+
+  return $ defModFI : join (sigs:decs)
 
+-------------------------------------------------------------------------------
+-- | Info about constructor being lifted; use 'mkCLInfo' to create one.
+data ConLiftInfo = CLInfo {
+    typeName :: Name
+  , funName  :: Name
+  , conName  :: Name
+  , effTArgs :: [Type]
+  , mName    :: Name
+  , resArg   :: Type
+  , conCxt   :: Cxt
+  , conArgs  :: [Type]
+  } deriving Show
+
+------------------------------------------------------------------------------
+-- | Creates info about constructor being lifted from parent type's and it's
+-- own info.
+mkCLInfo :: DatatypeInfo -> ConstructorInfo -> Q ConLiftInfo
+-- mkCLInfo dtInfo conInfo = do
+mkCLInfo DatatypeInfo {
+             datatypeName = typeName
+           , datatypeVars
+           , datatypeInstTypes
+           }
+         ConstructorInfo {
+             constructorName = conName
+           , constructorContext
+           , constructorFields
+           }
+  = do
+  let funName           = mapNameBase (mapHead toLower) conName
+      -- We extract equality constraints with variables to unify them
+      -- manually - this makes type errors more readable
+      (eqPairs, conCxt) = first (M.fromList . maybeAddResKind)
+                        $ partitionWith eqPairOrCxt constructorContext
+      -- We constrain kind of result to 'Type' if we get a type variable
+      -- TODO: we probably want to add requirement for KindSignatures into
+      -- documentation - PolyKinds are already mentioned...
+      -- TODO: This makes function signatures ugly - what is the best way to
+      -- restrict occurence of kind signature or completely remove it?
+      -- TODO: take result type by name or position?
+      maybeAddResKind   = maybe id ((:) . (, StarT))
```

Comment:
Too point-free. What's going on here?

Reply:
If we've got `Just` name of kind, put it in a pair with `StarT` for substitution and make it append to supplied list of substitutions, else leave list as-is.

Reply:
Some explicit lambdas would help here!

Reply:
`\case { Nothing -> id; Just rk -> ((rk, StarT):) }` is similarly terse but more obvious, and `LambdaCase` is already common in the codebase.

Reply:
```
      maybeAddResKind       = (\case Nothing -> id
                                     Just k  -> ((k, StarT) :))
                            $ tVarName $ tvKind $ last datatypeVars
```
looks fine :slightly_smiling_face:

Reply:
I'm happy with `maybe`, it's the `(:) . (, StarT)` but I don't want to have to unravel in my brain

Reply:
Hmm, what about `maybe id (\k ps -> (k, StarT) : ps)`?

Reply:
Much better; now I understand what's happening here!

Reply:
Great!